### PR TITLE
refactor(main): using registry mode build save images

### PR DESCRIPTION
**Pull request by cuisongliu**
### Summary
🚚🛠️📝

Refactored the registry and image saving logic to use the `crane` library and a temporary registry container. Moved and renamed some files and packages to improve the code structure and readability. Fixed a documentation error and updated the linter configuration.

> _To save images with sealos_
> _We refactored the registry code_
> _We used `crane` for auth and sync_
> _And moved some files to avoid a link_
> _We also fixed some docs and linter woe_

### Walkthrough
*  Update the linter configuration to skip the third-party code directory ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-6179837f7df53a6f05c522b6b7bb566d484d5465d9894fb04910dd08bb40dcc9L13-R13))
*  Fix the documentation link for the `toYaml` function ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-124d72db2503c78f1729137c07ec65494c3915b275c1766bf079d798adc5f279L44-R44))
*  Implement the image saving logic using the crane library and the temporary registry approach ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-8fecdf546a08e5fe09cd5106811fb7016c57d83b6873b81d9ac02903d996bcbdR25-R27), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-8fecdf546a08e5fe09cd5106811fb7016c57d83b6873b81d9ac02903d996bcbdL35), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-8fecdf546a08e5fe09cd5106811fb7016c57d83b6873b81d9ac02903d996bcbdL94-R100), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L16-R29), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L76-R75), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L101-R100), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-40842ce77ff674f441713045ba15b361515aef3c96aaa33e0d1f66e203b5ef1aR23-R24), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-40842ce77ff674f441713045ba15b361515aef3c96aaa33e0d1f66e203b5ef1aL28), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-40842ce77ff674f441713045ba15b361515aef3c96aaa33e0d1f66e203b5ef1aL45-R46), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-fa0d9dcb58aa3c0d341ff20196b1fcb75156b1bb10c2b659cd29393b36ca0bffL34-R38), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-fa0d9dcb58aa3c0d341ff20196b1fcb75156b1bb10c2b659cd29393b36ca0bffL42-R61), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-fa0d9dcb58aa3c0d341ff20196b1fcb75156b1bb10c2b659cd29393b36ca0bffL49-R68), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-fa0d9dcb58aa3c0d341ff20196b1fcb75156b1bb10c2b659cd29393b36ca0bffR75-R91))
*  Implement the registry synchronization logic using the crane library and the temporary registry approach ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L16-R29), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L76-R75), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-9d77d600601adda24842f567809994d01cf84d860574c5ed89288c53c3a5f7f5L101-R100))
*  Implement the registry interface using the crane library and the custom keychain ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-5c7fcb5d5095b8e5f7611ba601ff143825e1d615a0c32e73ecaea161d0007278L24-R29), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-5c7fcb5d5095b8e5f7611ba601ff143825e1d615a0c32e73ecaea161d0007278L103-R103), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-55f375c2b976f6fc3bedd745cf7de02ecaf5c2736370ddd702465993a368a9d2L24-R28), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-55f375c2b976f6fc3bedd745cf7de02ecaf5c2736370ddd702465993a368a9d2L48-R49), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-7fb966350558b15ebe16c2a67184c26ac4439bb6e59e23d2d3ca1491140d5fc5L15-R32), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-6dc9cfecffd5c2ad8108f95c656f81dc36033d61500a877e7d140e6b86490a20L17-R25), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-85690cb97863228b289bc0557e2434926a9352e77fd6140eeaaca86b47c4ba96L15-R15), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-0368094030a8ad8f7e5d73c8139e9619a5dd086867cd0536a410f23a123a748fL17-R17), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-0368094030a8ad8f7e5d73c8139e9619a5dd086867cd0536a410f23a123a748fR23), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-0368094030a8ad8f7e5d73c8139e9619a5dd086867cd0536a410f23a123a748fL28-R30), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-0368094030a8ad8f7e5d73c8139e9619a5dd086867cd0536a410f23a123a748fL36-R38), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-0368094030a8ad8f7e5d73c8139e9619a5dd086867cd0536a410f23a123a748fR73-R94), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-55ff3aacb53e97c6d79c56a5da5aee3a20319017941ae8224799a0f78d46b34fL17-R17), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-55ff3aacb53e97c6d79c56a5da5aee3a20319017941ae8224799a0f78d46b34fR23-R24), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-55ff3aacb53e97c6d79c56a5da5aee3a20319017941ae8224799a0f78d46b34fR84-R98))
*  Move the third-party code to a subdirectory and update the import paths ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-5279c1125c233c467c649b237d9834fc633ac73ce3c823c87f834cec0198ca1c), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-47d661b33de8fbf8ccd2e25602531c87dc95e2945489093b6458567305ac34be), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-02d9b76d3a2fbf2101987632eb7e0f012d19e1e7f988e3ed9783a4f7dcc43c09), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-46ff846e7e57e5d9cab5d4b34f3199314932d7d9801a6c11257e9e5f1e545326L27-R29), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-380262cb94ecaa3887b6e347bb7b72f2d20f677094a4d2ee58ae290bd69b41f2), [link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-2c34c9d332db820d77f0022b485310b27cc94ad533dd12914ade17b788741653L24-R24))
*  Delete the unused file `pkg/registry/commands/registry.go` ([link](https://github.com/labring/sealos/pull/3154/files?diff=unified&w=0#diff-480b4d7ce5bb933dd58f1adb881a43b5bd2ce8ceaaf8eb4fc4122f9f4be4ec9b))

### CODE REVIEW

File: pkg/registry/save/registry_save.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package save
+
+import (
+	"fmt"
+	"strings"
+	sync2 "sync"
```
Comment (fengxsong) [inline pkg/registry/save/registry_save.go:20]: stdsync

File: pkg/registry/sync/sync.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		return err
+	}
+	return retry.RetryIfNecessary(ctx, func() error {
+		_, err = copy.Image(ctx, policyContext, destRef, src, &copy.Options{
+			SourceCtx:          sys,
+			DestinationCtx:     sys,
+			ImageListSelection: selection,
+			ReportWriter:       os.Stdout,
+		})
+		if err != nil {
```
Comment (fengxsong) [inline pkg/registry/sync/sync.go:116]: return err

File: pkg/system/env.go [hunk 3]
```diff
@@ -96,6 +101,7 @@ const (
 	BuildahFormatConfigKey            = "BUILDAH_FORMAT"
 	ScpChecksumConfigKey              = "SCP_CHECKSUM"
 	RegistrySyncExperimentalConfigKey = "REGISTRY_SYNC_EXPERIMENTAL"
+	RegistrySaveExperimentalConfigKey = "REGISTRY_SAVE_EXPERIMENTAL"
```
Comment (fengxsong) [inline pkg/system/env.go:104]: they are all using `sync` actually, so consider using one config key `REGISTRY_SYNC_EXPERIMENTAL`?

Comment (cuisongliu) [reply pkg/system/env.go:104]: But the scenarios are different. One is to synchronize the registry image in the filesystem stage, and the other is to synchronize the registry image in the build stage.

File: pkg/registry/sync/sync.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			return err
+		}
+		for j := range refs {
+			destSuffix := strings.TrimPrefix(refs[j].DockerReference().String(), src)
+			destRef, err := docker.ParseReference(fmt.Sprintf("//%s", filepath.Join(dst, destSuffix)))
+			if err != nil {
+				return err
+			}
+			logger.Debug("copy image using registry sync mode: %s", destRef.DockerReference().String())
+			_, err = copy.Image(ctx, policyContext, destRef, refs[j], &copy.Options{
```
Comment (fengxsong) [inline pkg/registry/sync/sync.go:76]: add retry here.

**Review by bxy4543 — APPROVED** [review]

**Pull request merged**